Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp

```diff
@@ -17,22 +18,157 @@ void URobotVehicleMovementComponent::Initialize()
+                                                                       startPos,
+                                                                       endPos,
+                                                                       ECollisionChannel::ECC_Visibility,
+                                                                       TraceParams,
+                                                                       FCollisionResponseParams::DefaultResponseParam);
+                if (!IsFloorHit)
+                {
+                    // if no impact below, just consider this contact point is falling
+                    Hit.ImpactPoint = Contact->GetComponentLocation() - FVector(0., 0., FallingSpeed * InDeltaTime);
+                    Hit.Distance = RayOffsetUp + FallingSpeed * InDeltaTime;
+                }
+
+                // keep only the 3 contacts with shortest distances
+                if (NbContact < 3)
+                {
+                    Contacts[NbContact] = Hit.ImpactPoint;
+                    ContactsDistance[NbContact] = Hit.Distance;
+                    NbContact++;
+                }
+                else
+                {
+                    uint8 MaxContactDistanceIndex = 0;
+                    if (ContactsDistance[1] > ContactsDistance[0])
+                        MaxContactDistanceIndex = 1;
+                    if (ContactsDistance[2] > ContactsDistance[MaxContactDistanceIndex])
+                        MaxContactDistanceIndex = 2;
+                    if (Hit.Distance < ContactsDistance[MaxContactDistanceIndex])
+                    {
+                        ContactsDistance[MaxContactDistanceIndex] = Hit.Distance;
+                        Contacts[MaxContactDistanceIndex] = Hit.ImpactPoint;
+                    }
+                }
+            }
+
+            // get the normal vector of the plane going through these 3 points
+            FVector PlaneNormal = FVector::CrossProduct(Contacts[1] - Contacts[0], Contacts[2] - Contacts[0]);
+            PlaneNormal.Normalize(0.01);
+            if (PlaneNormal.Z < 0.)
+                PlaneNormal = -PlaneNormal;
+
+            FVector ForwardProjection = FVector::VectorPlaneProject(PawnOwner->GetActorForwardVector(), PlaneNormal);
+            PawnOwner->SetActorRotation(UKismetMathLibrary::MakeRotFromXZ(ForwardProjection, PlaneNormal));
+
+            if (MovingPlatform == nullptr)
+            {
+                // Moves the robot up or down, depending on impact position
+                float MinDistance = ContactsDistance[0];
+                if (ContactsDistance[1] < MinDistance)
+                    MinDistance = ContactsDistance[1];
+                if (ContactsDistance[2] < MinDistance)
+                    MinDistance = ContactsDistance[2];
+
+                MinDistance -= RayOffsetUp;
+
+                if (MinDistance > FallingSpeed * InDeltaTime)
+                    MinDistance = FallingSpeed * InDeltaTime;

```

Comment:
```suggestion
                 MinDistance = FMath::Min(MinDistance, FallingSpeed * InDeltaTime);
```
so the multiplication is done once only.